File: coding/src/com/algos/topics/li10_tree/bt/Bt_L404_SumOfLeftLeaves.java
package  com.algos.topics.li10_tree.bt;

import com.algos.topics.models.TreeNode;

public class Bt_L404_SumOfLeftLeaves {

	public static void main(String[] args) {
		Bt_L404_SumOfLeftLeaves obj = new Bt_L404_SumOfLeftLeaves();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		System.out.println(obj.sumOfLeftLeaves(root));

	}

	/*
	 * Problem: https://leetcode.com/problems/sum-of-left-leaves/
	 * Idea: As we need to differentiate between left and right, use a flag arg.
	 * */
	public int sumOfLeftLeaves(TreeNode root) {
        return sumOfLeftLeaves(root, false);
    }

    private int sumOfLeftLeaves(TreeNode root, boolean isLeft) {
        if(root == null)
			return 0;
        if(root.left == null && root.right == null)
            if(isLeft)
				return root.val;
            else
				return 0;

        return sumOfLeftLeaves(root.left, true) + sumOfLeftLeaves(root.right, false);
    }

}

File: dsa/topics/10_trees/bt/python/bt_l404__sum_of_left_leaves.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/sum-of-left-leaves/
# Idea: Use a flag to differentiate left and right child nodes.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def sum_of_left_leaves(root):
    def helper(node, is_left):
        if not node:
            return 0
        if not node.left and not node.right:  # leaf node
            return node.val if is_left else 0
        return helper(node.left, True) + helper(node.right, False)

    return helper(root, False)


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(sum_of_left_leaves(root))  # Expected: 7 (2 + 5)
